model/src/main/java/ru/training/karaf/repo/BookRepoImpl.java

```diff
@@ -0,0 +1,176 @@
+        try {
+            return Optional.of(em.createNamedQuery(BookDO.GET_BY_ID, BookDO.class).setParameter("id", id)
+                    .getSingleResult());
+        } catch (NoResultException e) {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public void changePrice(String name, String author, int price) {
+        Optional<BookDO> bookDO = template.txExpr(em -> getByNameAndAuthor(em, name, author));
+        if (bookDO.isPresent()) {
+            BookDO book = bookDO.get();
+            BookPriceRepoImpl bookPriceRepo = new BookPriceRepoImpl(template);
+            bookPriceRepo.create(price, book.getId());
+        } else {
+            RequestLogRepoImpl requestLogRepo = new RequestLogRepoImpl(template);
+            requestLogRepo.add(name, author, price);
+        }
+
+    }
+
+    @Override
+    public List<BookPrice> getPrices(long bookId) {
+        BookPriceRepoImpl repo = new BookPriceRepoImpl(template);
+        return repo.getByBook(bookId);
+    }
+
+    private Optional<BookDO> getByNameAndAuthor(EntityManager em, String name, String author) {try {
+        return Optional.of(em.createNamedQuery(BookDO.GET_BY_NAME_AND_AUTHOR, BookDO.class)
+                .setParameter("name", name)
+                .setParameter("author", author)
+                .getSingleResult());
+    } catch (NoResultException e) {
+        return Optional.empty();
+    }
+
+    }
+}
+
+class BookImpl implements Book {
+
+    private final BookDO bookDO;
+
+    BookImpl(BookDO bookDO) {
+        this.bookDO = bookDO;
+    }
+
+    BookDO getDO(){
+        return bookDO;
+    }
+
+    @Override
+    public long getId() {
+        return bookDO.getId();
+    }
+
+    @Override
+    public String getName() {
+        return bookDO.getName();
+    }
+
+    @Override
+    public String getUsr() {
+        return bookDO.getUsr().getLogin();
+    }
+
+    @Override
+    public BookDescription getBookDescription() {
+        BookDescriptionImpl desc = new BookDescriptionImpl(bookDO.getDescriptionDo());
+        //opt
+        return desc; //?
+    }
+
+    @Override
+    public Color getColor() {
+        return bookDO.getColor();
+    }
+
+    @Override
+    public Collection<String> getAuthorNames() {
+        return bookDO.getAuthors().stream().map(AuthorImpl::new).map(a -> a.getName()).collect(Collectors.toList());
+    }
+
+    @Override
+    public Collection<Author> getAuthors() {
+        return bookDO.getAuthors().stream().map(AuthorImpl::new).collect(Collectors.toList());
+    }
+    @Override
+    public <T> T unWrap(Class<T> clazz) {
+        if (clazz == BookDO.class) {
+            return (T) bookDO;
+        }
+        return null;
+        //throw new NoSuchFieldException("");
+    }
+
+    @Override
+    public int getPrice() {
+        Optional<BookPriceDO> bookPriceDO = bookDO.getPrice().stream().max(Comparator.comparing(o -> (int) o.getId()));
```

Comment:
Shouldn't we compare dates to get the latest price?

Also, are you sure we won't get LazyInitializationException here?